File: ts/src/app/list.ts
/**
 * List commands for the CLI.
 *
 * Each function takes a Storage (and optional config parameters) and returns
 * plain objects that can be JSON.stringify'd to match the Rust CLI output.
 */

import { readdir, readFile } from 'node:fs/promises';
import type { Dirent } from 'node:fs';
import path from 'node:path';
import toml from 'toml';
import { Decimal } from '../decimal.js';
import { type Storage } from '../storage/storage.js';
import { type ConnectionType } from '../models/connection.js';
import { type AccountType } from '../models/account.js';
import { type AssetType } from '../models/asset.js';
import { Id } from '../models/id.js';
import { MarketDataService } from '../market-data/service.js';
import { NullMarketDataStore, type MarketDataStore } from '../market-data/store.js';
import {
  formatDateYMD,
  formatRfc3339,
  formatRfc3339FromEpochNanos,
} from './format.js';
import { DEFAULT_IGNORE_CONFIG, type ResolvedConfig } from '../config.js';
import {
  applyTransactionAnnotationPatch,
  isEmptyTransactionAnnotation,
  type TransactionAnnotationType,
} from '../models/transaction-annotation.js';
import { valueInReportingCurrencyBestEffort } from './value.js';
import { compileTransactionIgnoreRules, shouldIgnoreTransaction } from './ignore-rules.js';
import {
  type ConnectionOutput,
  type AccountOutput,
  type BalanceOutput,
  type TransactionOutput,
  type PriceSourceOutput,
  type AllOutput,
} from './types.js';

function buildAccountsByConnection(accounts: AccountType[]): Map<string, Set<string>> {
  const accountsByConnection = new Map<string, Set<string>>();
  for (const account of accounts) {
    const connectionId = account.connection_id.asStr();
    const set = accountsByConnection.get(connectionId);
    if (set !== undefined) {
      set.add(account.id.asStr());
    } else {
      accountsByConnection.set(connectionId, new Set([account.id.asStr()]));
    }
  }
  return accountsByConnection;
}

function parseRfc3339ToEpochNanos(value: string): bigint | null {
  const m = value.match(
    /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,9}))?(Z|[+-]\d{2}:\d{2})$/,
  );
  if (m === null) {
    return null;
  }

  const year = Number.parseInt(m[1], 10);
  const month = Number.parseInt(m[2], 10);
  const day = Number.parseInt(m[3], 10);
  const hour = Number.parseInt(m[4], 10);
  const minute = Number.parseInt(m[5], 10);
  const second = Number.parseInt(m[6], 10);
  const fractional = m[7] ?? '';
  const tz = m[8];

  const fractionNanos = BigInt((fractional + '000000000').slice(0, 9));
  const localMs = Date.UTC(year, month - 1, day, hour, minute, second);

  let offsetMinutes = 0;
  if (tz !== 'Z') {
    const sign = tz.startsWith('-') ? -1 : 1;
    const tzHour = Number.parseInt(tz.slice(1, 3), 10);
    const tzMinute = Number.parseInt(tz.slice(4, 6), 10);
    offsetMinutes = sign * (tzHour * 60 + tzMinute);
  }

  const utcMs = localMs - offsetMinutes * 60 * 1000;
  return BigInt(utcMs) * 1000000n + fractionNanos;
}

function formatRfc3339PreservingRaw(date: Date, raw?: string): string {
  if (raw !== undefined) {
    const parsed = parseRfc3339ToEpochNanos(raw);
    if (parsed !== null) {
      return formatRfc3339FromEpochNanos(parsed.toString());
    }
  }
  return formatRfc3339(date);
}

function assetForListOutput(asset: AssetType): AssetType {
  switch (asset.type) {
    case 'currency':
      return { iso_code: asset.iso_code, type: 'currency' };
    case 'equity': {
      if (asset.exchange !== undefined) {
        return { exchange: asset.exchange, ticker: asset.ticker, type: 'equity' };
      }
      return { ticker: asset.ticker, type: 'equity' };
    }
    case 'crypto': {
      if (asset.network !== undefined) {
        return { network: asset.network, symbol: asset.symbol, type: 'crypto' };
      }
      return { symbol: asset.symbol, type: 'crypto' };
    }
  }
}

function resolveConnectionAccountIds(
  connection: ConnectionType,
  accounts: AccountType[],
  validIds: Set<string>,
): string[] {
  const result: string[] = [];
  const seen = new Set<string>();

  for (const accountId of connection.state.account_ids) {
    const accountIdStr = accountId.asStr();
    if (!validIds.has(accountIdStr)) continue;
    if (seen.has(accountIdStr)) continue;
    seen.add(accountIdStr);
    result.push(accountIdStr);
  }

  for (const account of accounts) {
    if (!account.connection_id.equals(connection.state.id)) continue;
    const accountIdStr = account.id.asStr();
    if (!validIds.has(accountIdStr)) continue;
    if (seen.has(accountIdStr)) continue;
    seen.add(accountIdStr);
    result.push(accountIdStr);
  }

  return result;
}

// ---------------------------------------------------------------------------
// listConnections
// ---------------------------------------------------------------------------

/**
 * List all connections with account counts and formatted timestamps.
 */
export async function listConnections(storage: Storage): Promise<ConnectionOutput[]> {
  const [connections, accounts] = await Promise.all([
    storage.listConnections(),
    storage.listAccounts(),
  ]);
  const accountsByConnection = buildAccountsByConnection(accounts);

  return connections.map((conn) => {
    const validIds = accountsByConnection.get(conn.state.id.asStr()) ?? new Set<string>();
    const accountIds = resolveConnectionAccountIds(conn, accounts, validIds);

    const last_sync = conn.state.last_sync
      ? formatRfc3339PreservingRaw(conn.state.last_sync.at, conn.state.last_sync.at_raw)
      : null;

    return {
      id: conn.state.id.asStr(),
      name: conn.config.name,
      synchronizer: conn.config.synchronizer,
      status: conn.state.status,
      account_count: accountIds.length,
      last_sync,
    };
  });
}

// ---------------------------------------------------------------------------
// listAccounts
// ---------------------------------------------------------------------------

/**
 * List all accounts with basic fields.
 */
export async function listAccounts(storage: Storage): Promise<AccountOutput[]> {
  const accounts = await storage.listAccounts();
  return accounts.map((a) => ({
    id: a.id.asStr(),
    name: a.name,
    connection_id: a.connection_id.asStr(),
    tags: [...a.tags],
    active: a.active,
  }));
}

// ---------------------------------------------------------------------------
// listBalances
// ---------------------------------------------------------------------------

/**
 * List latest balances for all accounts.
 *
 * `value_in_reporting_currency` mirrors Rust behavior:
 * - same-currency amounts are normalized and returned
 * - equities/crypto use cached close prices and may fall back to same-day quotes
 *   when close is missing (and FX when needed)
 * - missing price/FX data returns `null`
 */
export async function listBalances(
  storage: Storage,
  config: ResolvedConfig,
  marketDataStore?: MarketDataStore,
): Promise<BalanceOutput[]> {
  const [connections, accounts] = await Promise.all([
    storage.listConnections(),
    storage.listAccounts(),
  ]);
  const accountsByConnection = buildAccountsByConnection(accounts);
  const reportingCurrencyUpper = config.reporting_currency.trim().toUpperCase();
  const currencyDecimals = config.display.currency_decimals;
  const marketData = new MarketDataService(marketDataStore ?? new NullMarketDataStore());
  const result: BalanceOutput[] = [];

  for (const connection of connections) {
    const validIds = accountsByConnection.get(connection.state.id.asStr()) ?? new Set<string>();
    const accountIds = resolveConnectionAccountIds(connection, accounts, validIds);

    for (const accountId of accountIds) {
      const snapshot = await storage.getLatestBalanceSnapshot(Id.fromString(accountId));
      if (!snapshot) continue;

      const asOfDate = formatDateYMD(snapshot.timestamp);
      for (const balance of snapshot.balances) {
        const valueInReportingCurrencyValue = await valueInReportingCurrencyBestEffort(
          marketData,
          balance.asset,
          balance.amount,
          reportingCurrencyUpper,
          asOfDate,
          currencyDecimals,
        );

        result.push({
          account_id: accountId,
          asset: assetForListOutput(balance.asset),
          amount: balance.amount,
          value_in_reporting_currency: valueInReportingCurrencyValue,
          reporting_currency: reportingCurrencyUpper,
          timestamp: formatRfc3339PreservingRaw(snapshot.timestamp, snapshot.timestamp_raw),
        });
      }
    }
  }

  return result;
}

// ---------------------------------------------------------------------------
// listTransactions
// ---------------------------------------------------------------------------

/**
 * List all transactions for all accounts, filtered by date range.
 */
export async function listTransactions(
  storage: Storage,
  startStr?: string,
  endStr?: string,
  config?: ResolvedConfig,
  sortByAmount = false,
  skipIgnored = true,
): Promise<TransactionOutput[]> {
  const endDate = endStr ? new Date(endStr + 'T23:59:59.999Z') : new Date();
  const startDate = startStr
    ? new Date(startStr + 'T00:00:00Z')
    : new Date(endDate.getTime() - 30 * 24 * 60 * 60 * 1000);

  const [accounts, connections] = await Promise.all([
    storage.listAccounts(),
    storage.listConnections(),
  ]);
  const connectionsById = new Map(connections.map((connection) => [connection.state.id.asStr(), connection]));
  const ignoreRules = skipIgnored
    ? compileTransactionIgnoreRules(config?.ignore ?? DEFAULT_IGNORE_CONFIG)
    : [];
  const result: TransactionOutput[] = [];

  for (const account of accounts) {
    const connection = connectionsById.get(account.connection_id.asStr());
    const connection_id = account.connection_id.asStr();
    const connection_name = connection?.config.name ?? '';
    const synchronizer = connection?.config.synchronizer ?? '';

    const transactions = await storage.getTransactions(account.id);
    const patches = await storage.getTransactionAnnotationPatches(account.id);

    // Materialize last-write-wins annotation state per transaction id.
    const annByTx = new Map<string, TransactionAnnotationType>();
    for (const p of patches) {
      const key = p.transaction_id.asStr();
      const base = annByTx.get(key) ?? { transaction_id: p.transaction_id };
      annByTx.set(key, applyTransactionAnnotationPatch(base, p));
    }

    for (const tx of transactions) {
      if (tx.timestamp < startDate || tx.timestamp > endDate) continue;
      if (
        skipIgnored &&
        shouldIgnoreTransaction(ignoreRules, {
          account_id: account.id.asStr(),
          account_name: account.name,
          connection_id,
          connection_name,
          synchronizer,
          description: tx.description,
          status: tx.status,
          amount: tx.amount,
        })
      ) {
        continue;
      }

      const ann = annByTx.get(tx.id.asStr());
      const annotation =
        ann && !isEmptyTransactionAnnotation(ann)
          ? {
              ...(ann.description !== undefined ? { description: ann.description } : {}),
              ...(ann.note !== undefined ? { note: ann.note } : {}),
              ...(ann.category !== undefined ? { category: ann.category } : {}),
              ...(ann.tags !== undefined ? { tags: ann.tags } : {}),
            }
          : undefined;

      const out: TransactionOutput = {
        id: tx.id.asStr(),
        account_id: account.id.asStr(),
        account_name: account.name,
        timestamp: formatRfc3339PreservingRaw(tx.timestamp, tx.timestamp_raw),
        description: tx.description,
        amount: tx.amount,
        asset: assetForListOutput(tx.asset),
        status: tx.status,
      };
      if (annotation !== undefined) {
        out.annotation = annotation;
      }
      result.push(out);
    }
  }

  if (sortByAmount) {
    result.sort((a, b) => {
      let left: Decimal | null = null;
      let right: Decimal | null = null;
      try {
        left = new Decimal(a.amount);
      } catch {
        left = null;
      }
      try {
        right = new Decimal(b.amount);
      } catch {
        right = null;
      }

      if (left !== null && right !== null) return left.comparedTo(right);
      if (left === null && right !== null) return 1;
      if (left !== null && right === null) return -1;
      return a.amount.localeCompare(b.amount);
    });
  }

  return result;
}

// ---------------------------------------------------------------------------
// listPriceSources
// ---------------------------------------------------------------------------

const KNOWN_SOURCE_TYPES = new Set<string>([
  'eodhd',
  'twelve_data',
  'alpha_vantage',
  'marketstack',
  'coingecko',
  'cryptocompare',
  'coincap',
  'frankfurter',
]);

function toOutputSourceType(value: string): string | null {
  const normalized = value.trim().toLowerCase();
  if (!KNOWN_SOURCE_TYPES.has(normalized)) return null;
  // Rust currently formats source type with Debug + lowercase, so underscores are removed.
  return normalized.replaceAll('_', '');
}

function parsePriority(value: unknown): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) return 100;
  return Math.trunc(value);
}

type PriceSourceToml = {
  type?: unknown;
  enabled?: unknown;
  priority?: unknown;
  credentials?: unknown;
};

/**
 * List configured price sources from `data_dir/price_sources/<name>/source.toml`.
 *
 * Mirrors Rust behavior:
 * - include only enabled sources (enabled defaults to true)
 * - skip invalid configs
 * - sort by ascending priority
 */
export async function listPriceSources(dataDir?: string): Promise<PriceSourceOutput[]> {
  if (dataDir === undefined) return [];

  const sourcesDir = path.join(dataDir, 'price_sources');
  let entries: Dirent<string>[];
  try {
    entries = await readdir(sourcesDir, { withFileTypes: true, encoding: 'utf8' });
  } catch (err: unknown) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
      return [];
    }
    throw err;
  }

  const output: PriceSourceOutput[] = [];

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const sourceTomlPath = path.join(sourcesDir, entry.name, 'source.toml');
    let content: string;
    try {
      content = await readFile(sourceTomlPath, 'utf8');
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code === 'ENOENT') continue;
      continue;
    }

    let parsed: PriceSourceToml;
    try {
      parsed = toml.parse(content) as PriceSourceToml;
    } catch {
      continue;
    }

    if (typeof parsed.type !== 'string') continue;
    const outputType = toOutputSourceType(parsed.type);
    if (outputType === null) continue;

    const enabled = typeof parsed.enabled === 'boolean' ? parsed.enabled : true;
    if (!enabled) continue;

    output.push({
      name: entry.name,
      type: outputType,
      enabled,
      priority: parsePriority(parsed.priority),
      has_credentials: parsed.credentials !== undefined && parsed.credentials !== null,
    });
  }

  output.sort((a, b) => a.priority - b.priority);
  return output;
}

// ---------------------------------------------------------------------------
// listAll
// ---------------------------------------------------------------------------

/**
 * Combine all list outputs into a single object.
 */
export async function listAll(
  storage: Storage,
  config: ResolvedConfig,
  marketDataStore?: MarketDataStore,
): Promise<AllOutput> {
  const [connections, accounts, balances, priceSources] = await Promise.all([
    listConnections(storage),
    listAccounts(storage),
    listBalances(storage, config, marketDataStore),
    listPriceSources(config.data_dir),
  ]);

  return {
    connections,
    accounts,
    price_sources: priceSources,
    balances,
  };
}

File: src/app/list.rs
use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::str::FromStr;

use anyhow::{Context, Result};
use chrono::{NaiveDate, Utc};
use chrono_tz::Tz;
use rust_decimal::Decimal;

use crate::config::{DisplayConfig, ResolvedConfig};
use crate::format::{currency_symbol, format_base_currency_display};
use crate::market_data::{MarketDataServiceBuilder, PriceSourceRegistry};
use crate::models::{Asset, Id, TransactionAnnotation};
use crate::storage::Storage;

use super::classification::{
    effective_transaction_subtags, effective_transaction_tags, provider_virtual_tag_hierarchy,
};
use super::ignore_rules::{TransactionIgnoreInput, TransactionIgnoreMatcher};
use super::value::value_in_reporting_currency_best_effort;
use super::{
    AccountOutput, AllOutput, BalanceOutput, ConnectionOutput, PriceSourceOutput,
    TransactionAnnotationOutput, TransactionOutput,
};

#[derive(Debug, Clone)]
enum TransactionDateTz {
    Utc,
    Local,
    Named(Tz),
}

impl TransactionDateTz {
    fn parse(tz: Option<String>) -> Result<Self> {
        let Some(tz) = tz else {
            return Ok(Self::Utc);
        };
        let trimmed = tz.trim();
        if trimmed.is_empty() || trimmed.eq_ignore_ascii_case("utc") {
            return Ok(Self::Utc);
        }
        if trimmed.eq_ignore_ascii_case("local") || trimmed.eq_ignore_ascii_case("current") {
            return Ok(Self::Local);
        }
        let named = trimmed.parse::<Tz>().with_context(|| {
            format!("Invalid timezone '{trimmed}' (expected IANA name, e.g. America/New_York)")
        })?;
        Ok(Self::Named(named))
    }

    fn date_for(&self, timestamp: chrono::DateTime<Utc>) -> NaiveDate {
        match self {
            Self::Utc => timestamp.date_naive(),
            Self::Local => timestamp.with_timezone(&chrono::Local).date_naive(),
            Self::Named(tz) => timestamp.with_timezone(tz).date_naive(),
        }
    }
}

fn annotation_ignores_spending(annotation: &TransactionAnnotation) -> bool {
    annotation.ignores_spending()
}

pub async fn list_connections(storage: &dyn Storage) -> Result<Vec<ConnectionOutput>> {
    let connections = storage.list_connections().await?;
    let accounts = storage.list_accounts().await?;
    let mut accounts_by_connection: HashMap<Id, HashSet<Id>> = HashMap::new();
    for account in accounts {
        accounts_by_connection
            .entry(account.connection_id.clone())
            .or_default()
            .insert(account.id.clone());
    }
    let mut output = Vec::new();

    for c in connections {
        let valid_ids = accounts_by_connection
            .get(c.id())
            .cloned()
            .unwrap_or_default();
        let mut account_ids: HashSet<Id> = c
            .state
            .account_ids
            .iter()
            .filter(|id| valid_ids.contains(*id))
            .cloned()
            .collect();
        for account_id in valid_ids {
            account_ids.insert(account_id);
        }

        output.push(ConnectionOutput {
            id: c.id().to_string(),
            name: c.config.name.clone(),
            synchronizer: c.config.synchronizer.clone(),
            status: c.state.status.to_string(),
            account_count: account_ids.len(),
            last_sync: c.state.last_sync.as_ref().map(|ls| ls.at.to_rfc3339()),
        });
    }

    Ok(output)
}

pub async fn list_accounts(storage: &dyn Storage) -> Result<Vec<AccountOutput>> {
    let accounts = storage.list_accounts().await?;
    let mut output = Vec::new();

    for a in accounts {
        let exclude_from_portfolio = storage
            .get_account_config(&a.id)?
            .and_then(|config| config.exclude_from_portfolio)
            .unwrap_or(false);

        output.push(AccountOutput {
            id: a.id.to_string(),
            name: a.name.clone(),
            connection_id: a.connection_id.to_string(),
            tags: a.tags.clone(),
            active: a.active,
            exclude_from_portfolio,
        });
    }

    Ok(output)
}

pub fn list_price_sources(data_dir: &Path) -> Result<Vec<PriceSourceOutput>> {
    let mut registry = PriceSourceRegistry::new(data_dir);
    registry.load()?;

    let mut output = Vec::new();
    for s in registry.sources() {
        output.push(PriceSourceOutput {
            name: s.name.clone(),
            source_type: format!("{:?}", s.config.source_type).to_lowercase(),
            enabled: s.config.enabled,
            priority: s.config.priority,
            has_credentials: s.config.credentials.is_some(),
        });
    }

    Ok(output)
}

fn format_currency_amount_display(
    amount: &str,
    currency: &str,
    display: &DisplayConfig,
    symbol_override: Option<&str>,
) -> Option<String> {
    let value = Decimal::from_str(amount).ok()?;
    let currency = currency.trim().to_uppercase();
    let symbol = symbol_override.or_else(|| currency_symbol(&currency));
    let formatted = format_base_currency_display(
        value,
        display.currency_decimals,
        display.currency_grouping,
        symbol,
        display.currency_fixed_decimals,
    );
    Some(if symbol.is_some() {
        formatted
    } else {
        format!("{formatted} {currency}")
    })
}

pub async fn list_balances(
    storage: &dyn Storage,
    config: &ResolvedConfig,
) -> Result<Vec<BalanceOutput>> {
    let market_data = MarketDataServiceBuilder::for_data_dir(&config.data_dir)
        .with_quote_staleness(config.refresh.price_staleness)
        .build()
        .await;

    let connections = storage.list_connections().await?;
    let accounts = storage.list_accounts().await?;
    let mut accounts_by_connection: HashMap<Id, HashSet<Id>> = HashMap::new();
    for account in accounts {
        accounts_by_connection
            .entry(account.connection_id.clone())
            .or_default()
            .insert(account.id);
    }
    let mut output = Vec::new();

    for conn in connections {
        let valid_ids = accounts_by_connection
            .get(conn.id())
            .cloned()
            .unwrap_or_default();
        let mut account_ids = Vec::new();
        let mut seen_ids: HashSet<Id> = HashSet::new();
        for account_id in &conn.state.account_ids {
            if !valid_ids.contains(account_id) {
                continue;
            }
            if seen_ids.insert(account_id.clone()) {
                account_ids.push(account_id.clone());
            }
        }
        for account_id in valid_ids {
            if seen_ids.insert(account_id.clone()) {
                account_ids.push(account_id);
            }
        }

        for account_id in &account_ids {
            if let Some(snapshot) = storage.get_latest_balance_snapshot(account_id).await? {
                for balance in snapshot.balances {
                    let value_in_reporting_currency = value_in_reporting_currency_best_effort(
                        &market_data,
                        &balance.asset,
                        &balance.amount,
                        &config.reporting_currency,
                        snapshot.timestamp.date_naive(),
                        config.display.currency_decimals,
                    )
                    .await?;

                    let reporting_currency = config.reporting_currency.to_uppercase();
                    let reporting_currency_symbol = config
                        .display
                        .currency_symbol
                        .as_deref()
                        .or_else(|| currency_symbol(&reporting_currency));
                    let amount_display = match &balance.asset {
                        Asset::Currency { iso_code } => {
                            let symbol_override =
                                if iso_code.eq_ignore_ascii_case(&reporting_currency) {
                                    reporting_currency_symbol
                                } else {
                                    None
                                };
                            format_currency_amount_display(
                                &balance.amount,
                                iso_code,
                                &config.display,
                                symbol_override,
                            )
                        }
                        Asset::ManualValue { currency, .. } => {
                            let symbol_override =
                                if currency.eq_ignore_ascii_case(&reporting_currency) {
                                    reporting_currency_symbol
                                } else {
                                    None
                                };
                            format_currency_amount_display(
                                &balance.amount,
                                currency,
                                &config.display,
                                symbol_override,
                            )
                        }
                        Asset::Equity { .. } | Asset::Crypto { .. } => None,
                    };
                    let value_in_reporting_currency_display =
                        value_in_reporting_currency.as_deref().and_then(|value| {
                            format_currency_amount_display(
                                value,
                                &reporting_currency,
                                &config.display,
                                reporting_currency_symbol,
                            )
                        });

                    output.push(BalanceOutput {
                        account_id: account_id.to_string(),
                        asset: serde_json::to_value(&balance.asset)?,
                        amount: balance.amount,
                        amount_display,
                        cost_basis: balance.cost_basis,
                        value_in_reporting_currency,
                        value_in_reporting_currency_display,
                        reporting_currency,
                        reporting_currency_symbol: reporting_currency_symbol.map(str::to_string),
                        timestamp: snapshot.timestamp.to_rfc3339(),
                    });
                }
            }
        }
    }

    Ok(output)
}

pub async fn list_transactions(
    storage: &dyn Storage,
    start: Option<String>,
    end: Option<String>,
    tz: Option<String>,
    sort_by_amount: bool,
    skip_ignored: bool,
    config: &ResolvedConfig,
) -> Result<Vec<TransactionOutput>> {
    let date_tz = TransactionDateTz::parse(tz)?;
    let end_date = match &end {
        Some(s) => NaiveDate::parse_from_str(s, "%Y-%m-%d")
            .with_context(|| format!("Invalid end date: {s}"))?,
        None => date_tz.date_for(Utc::now()),
    };
    let start_date = match &start {
        Some(s) => NaiveDate::parse_from_str(s, "%Y-%m-%d")
            .with_context(|| format!("Invalid start date: {s}"))?,
        None => end_date - chrono::Duration::days(30),
    };

    let ignore_matcher = if skip_ignored {
        Some(TransactionIgnoreMatcher::from_configs(
            &config.ignore,
            &config.spending,
        )?)
    } else {
        None
    };
    let accounts = storage.list_accounts().await?;
    let connections = storage.list_connections().await?;
    let connections_by_id: HashMap<String, crate::models::Connection> = connections
        .into_iter()
        .map(|c| (c.id().to_string(), c))
        .collect();
    let mut output = Vec::new();
    let ignored_account_tags: HashSet<String> = if skip_ignored {
        config
            .spending
            .ignore_tags
            .iter()
            .filter_map(|tag| {
                let trimmed = tag.trim();
                if trimmed.is_empty() {
                    None
                } else {
                    Some(trimmed.to_lowercase())
                }
            })
            .collect()
    } else {
        HashSet::new()
    };

    for account in accounts {
        if skip_ignored
            && !ignored_account_tags.is_empty()
            && account.tags.iter().any(|tag| {
                let trimmed = tag.trim();
                !trimmed.is_empty() && ignored_account_tags.contains(&trimmed.to_lowercase())
            })
        {
            continue;
        }

        let connection = connections_by_id.get(&account.connection_id.to_string());
        let connection_id = account.connection_id.to_string();
        let connection_name = connection
            .map(|c| c.config.name.as_str())
            .unwrap_or_default();
        let synchronizer = connection
            .map(|c| c.config.synchronizer.as_str())
            .unwrap_or_default();

        let transactions = storage.get_transactions(&account.id).await?;
        let patches = storage
            .get_transaction_annotation_patches(&account.id)
            .await?;

        // Materialize last-write-wins annotation state per transaction id.
        let mut annotations_by_tx: HashMap<Id, TransactionAnnotation> = HashMap::new();
        for patch in patches {
            let tx_id = patch.transaction_id.clone();
            let ann = annotations_by_tx
                .entry(tx_id.clone())
                .or_insert_with(|| TransactionAnnotation::new(tx_id));
            patch.apply_to(ann);
        }

        for tx in transactions {
            let ann = annotations_by_tx.get(&tx.id);
            let tx_date = ann
                .and_then(|annotation| annotation.effective_date)
                .unwrap_or_else(|| date_tz.date_for(tx.timestamp));
            if tx_date < start_date || tx_date > end_date {
                continue;
            }
            let status = format!("{:?}", tx.status).to_lowercase();

            if skip_ignored {
                if tx
                    .standardized_metadata
                    .as_ref()
                    .and_then(|md| md.is_internal_transfer_hint)
                    .unwrap_or(false)
                {
                    continue;
                }
                if ignore_matcher.as_ref().is_some_and(|matcher| {
                    matcher.is_match(&TransactionIgnoreInput {
                        account_id: account.id.as_str(),
                        account_name: &account.name,
                        connection_id: &connection_id,
                        connection_name,
                        synchronizer,
                        description: &tx.description,
                        status: &status,
                        amount: &tx.amount,
                    })
                }) {
                    continue;
                }
            }

            let annotation = annotations_by_tx.get(&tx.id).and_then(|ann| {
                if ann.is_empty() {
                    None
                } else {
                    Some(TransactionAnnotationOutput {
                        description: ann.description.clone(),
                        note: ann.note.clone(),
                        tags: ann.tags.clone(),
                        subtags: ann.subtags.clone(),
                        effective_date: ann.effective_date.map(|d| d.to_string()),
                        ignore_spending: ann.ignore_spending,
                    })
                }
            });
            let raw_annotation = annotations_by_tx.get(&tx.id);
            let provider_hierarchy = provider_virtual_tag_hierarchy(
                tx.standardized_metadata.as_ref(),
                &tx.synchronizer_data,
                &config.tags,
            );
            let tags =
                effective_transaction_tags(raw_annotation, &provider_hierarchy, &config.tags);
            let subtags =
                effective_transaction_subtags(raw_annotation, &provider_hierarchy, &config.tags);
            if skip_ignored
                && annotations_by_tx
                    .get(&tx.id)
                    .is_some_and(annotation_ignores_spending)
            {
                continue;
            }

            output.push(TransactionOutput {
                id: tx.id.to_string(),
                account_id: account.id.to_string(),
                account_name: account.name.clone(),
                timestamp: tx.timestamp.to_rfc3339(),
                description: tx.description.clone(),
                amount: tx.amount.clone(),
                asset: serde_json::to_value(&tx.asset).unwrap_or_default(),
                status,
                tags,
                subtags,
                annotation,
                standardized_metadata: tx.standardized_metadata.clone(),
            });
        }
    }

    if sort_by_amount {
        output.sort_by(|a, b| {
            let left = Decimal::from_str(&a.amount);
            let right = Decimal::from_str(&b.amount);
            match (left, right) {
                (Ok(la), Ok(rb)) => la.cmp(&rb),
                (Err(_), Ok(_)) => std::cmp::Ordering::Greater,
                (Ok(_), Err(_)) => std::cmp::Ordering::Less,
                (Err(_), Err(_)) => a.amount.cmp(&b.amount),
            }
        });
    }

    Ok(output)
}

pub async fn list_all(storage: &dyn Storage, config: &ResolvedConfig) -> Result<AllOutput> {
    Ok(AllOutput {
        connections: list_connections(storage).await?,
        accounts: list_accounts(storage).await?,
        price_sources: list_price_sources(&config.data_dir)?,
        balances: list_balances(storage, config).await?,
    })
}

#[cfg(test)]
#[path = "../../tests/unit/app/list_tests.rs"]
mod list_tests;
